## Библиотека `NumPy`: массивы `NumPy`, их отличие от списков и преимущества по сравнению с последовательностями в `Python`.

В ходе предыдущего занятия вы познакомились с последовательностями – типами данных, позволяющими хранить множество различных значений произвольных типов. При численном решении физических задач вам чаще всего придётся использовать списки, поскольку они являются изменяемым (mutable) типом данных, позволяющим не только хранить упорядоченные последовательности элементов, но и удобно оперировать ими.

Наряду с перечисленными преимуществами, тип данных `list` обладает и существенным ограничением – низкой скоростью обработки данных, сужающим возможности его применения при решении задач, требующих работы с большими объёмами информации. Это ограничение связано с тем, что элементами списка могут быть различные типы данных, так что интерпретатору при обращении к каждому элементу списка требуется дополнительное время, в т. ч. для проверки его типа.

Для решения этой проблемы в `Python` была разработана библиотека (или модуль, если говорить в терминах `Python`) **`NumPy`**. С понятием модуля и возможностями его использования мы будем знакомиться детальнее в рамках лекции 8. Ниже мы рассмотрим основной функционал модуля `NumPy`, который может быть использован при решении задач.

## Массивы `NumPy`. Объявление массивов, заполнение с использованием цикла. Заполнение случайными числами. Извлечение из массива элемента с заданным номером.

Пакет `NumPy` определяет новый тип данных – $N$-мерный массив (`ndarray`). Массив – это упорядоченный набор элементов для хранения данных *одного* типа, идентифицируемых с помощью одного или нескольких индексов. 

Благодаря тому, что тип элементов массива `NumPy` един, интерпертатору нет необходимости проверять его каждый раз при обращении к тому или иному элементу, что значительно сокращает время на обработку данных.

### Объявление массивов

Для создания массива необходимо подключить модуль `numpy` (так, как это показано на примере ниже) и вызвать функцию `array`. Функция принимает в качестве аргумента любую последовательность (список, кортеж и т. д.) и преобразует его в массив `ndarray`. При этом важно следить за тем, чтобы все элементы последовательности имели один и тот же тип!

In [ ]:
import math  # подключаем модуль Math – для работы с математическими функциями
import numpy as np  # "np" – это псевдоним, сокращенное имя модуля numpy

# преобразуем список списков в массив ndarray
a = np.array([[1, math.e, math.pi], [4, 5, 6]])

print(a, end='\n\n')  # выводим на печать элементы массива,
print(a.shape)        # его размерность
print(a.dtype)        # и тип элементов
print()

# попробуем сделать массив-компот из данных разных типов
compote = np.array([[1e0, 'e', 'Pi'], [(4), 'five', 6]])
print(compote)
print(compote.dtype)
print(type(compote))
print(type(compote[1, 0]))


[[1.         2.71828183 3.14159265]
 [4.         5.         6.        ]]

(2, 3)
float64

[['1.0' 'e' 'Pi']
 ['4' 'five' '6']]
<U32
<class 'numpy.ndarray'>
<class 'numpy.str_'>


Как видите, массив-компот все же удалось создать, но все его элементы превратились в строки (метод `dtype` выдал "U32", что означает строковые данные в *unicode*).

Обратите внимание на то, что созданный массив `a` обладает рядом атрибутов.

* Атрибут `shape` (буквально, "форма") – кортеж, в котором хранятся размерности массива, – число элементов вдоль каждой оси массива (строки и столбцы в соответствующем порядке).
* Атрибут `dtype` (сокращенно от data type) – возвращает тип элементов, которые хранятся в массиве.

Общая рекомендация при работе с массивами: лучше всегда в явном виде указывать тип данных, которые вы хотите поместить в этот массив. Это поможет избежать трудно отслеживаемых опечаток, приводящих к преобразованию данных и потере точности. Таким образом, если бы мы записали массив `compote` следующим образом:
~~~python
compote = np.array([[1e0, 'e', 'Pi'], [(4), 'five', 6]], dtype=float)
~~~
то интерпретатор выдал бы ошибку.

Полный список атрибутов, как и в рассмотренных ранее примерах, можно получить, вызвав команду:
~~~python
dir(a)
~~~
или
~~~python
help(np)  # работает после импортирования модуля
~~~

Также отметим то, как массив выводится на печать: внешне он выглядит как "список списков": внешняя пара квадратных скобок ограничивает весь массив, а каждая строка в нем заключена в свою пару квадратных скобок. При этом элементы в каждой строке выстроены так, чтобы визуально отделить столбцы (в вывод включены символы табуляции `'\t'`).

##### Два способа задать массив
Функция `np.array()` позволяет создавать массив из уже существующей последовательности (иными словами, уже **заполненный данными**). Для создания массива, элементы которого не инициализированы в начальный момент (**пустой массив**), необходимо использовать функцию `np.ndarray()`. Обязательный аргумент этой функции – размерность массива – задается кортежом значений.

Рассмотрим пример:

In [ ]:
import numpy as np  # подключаем NumPy

a = np.ndarray((3, 4, 5))  # создаём трехмерный массив: 3 x 4 x 5 элементов
b = np.ndarray(a.shape, dtype = float)  # создаём массив по форме массива "a"

# выводим содержимое массивов на печать
print(a)
print('\n' + '-'*70 + '\n')  # в качестве визуального разделителя между массивами
print(b)


[[[4.65570908e-310 0.00000000e+000 2.10471965e-321             nan
   1.12646967e-321]
  [0.00000000e+000 1.41043067e+190 5.29874766e+180 4.27255602e+180
   3.99550968e+252]
  [5.83439496e+252 9.08254283e+223 7.19508849e+159 1.10255763e+214
   2.66061854e-260]
  [1.70555268e+161 3.10349169e+169 5.07324776e-119 6.01334626e-154
   7.28021853e+199]]

 [[3.80985005e+180 9.10016856e+276 2.45905750e-154 1.78711409e+161
   5.84938624e+199]
  [4.25551644e+228 4.56626968e+257 3.56405265e-057 3.65881510e+233
   2.71008188e+155]
  [5.81564546e+252 1.21697542e-152 2.19527118e-152 6.81454101e-067
   3.97977870e-062]
  [1.04803181e-042 9.77093792e-153 4.63266893e-066 1.07766100e-075
   5.59693173e-042]]

 [[1.73932301e+156 1.01766150e-028 9.00120641e+223 2.17162124e+214
   2.08460545e-115]
  [1.13443847e-036 1.08785532e+155 7.28197344e+223 1.41043076e+190
   5.29874766e+180]
  [2.04736806e+190 8.56232557e+194 2.43580036e-152 1.37800895e-094
   9.75022342e+199]
  [1.49005471e+195 6.96764527e+252 2.63

Обратите внимание, что созданные при помощи команды `np.ndarray()` массивы можно вывести на печать, однако в их элементах хранится "абракадабра" – случайный набор значений, величина которых близка к нулю или к бесконечности (а в массиве `a` даже встречается значение `nan` – *not a number* – не число).

С точки зрения интерпретации массивов в двумерном представлении (как набор строк и столбцов) можно считать, что приведенные выше массивы содержат в себе по 3 "слоя" (первый индекс в размерности), каждый из которых представляет собой 4 строки (второй индекс) и 5 столбцов (третий). Каждый из "слоев" при выводе массива на печать оказывается автоматически отделен от следующего пустой строкой – для удобства чтения.

### Обращение к элементам массива

Итак, мы создали ряд массивов. Для того, чтобы обратиться к какому-либо элементу массива, используется тот же способ, что и в случае последовательностей: после указания имени массива необходимо в квадратных скобках ввести индексы нужного элемента (отсчитываемые от нуля, как и прежде).

Индексы указываются внутри одной пары скобкок через запятую:

~~~python
a[2, 1, 0]
~~~

– в отличие от многомерной последовательности данных, задаваемой, например, с помощью вложенных списков. Для сравнения: если бы массив `a` был реализован именно как список, обращение к его элементам происходило бы так:

~~~python
a_list[2][1][0]
~~~

Для хранения и создания подобного объекта потребовалось бы использование трёхкратно (!) вложенных списков, что гораздо сложнее создания массива как с точки зрения реализации кода, так и с вычислительной позиции (упомянутая выше скорость обработки данных). При работе с многомерными данными, размерность которых превышает 3, сложность подобной задачи стремительно увеличивается!

Вернёмся к массивам `NumPy`. Поскольку тип `ndarray` – **изменямый** тип данных, то, обращаясь к элементам массива, можно не только получать информацию о них, но и менять значения, которые в них хранятся, например:

~~~python
a[2, 1, 0] = 5
~~~

##### Срезы
Ещё одной важной особенностью синтаксиса массивов `NumPy` является возможность использования срезов при работе с индексами массива.
![title](image/matrica.png)

Сравните:

In [ ]:
# работа со срезами
import numpy as np

# создадим новый пробный массив "a"
a = np.reshape(np.linspace(1, 12, 12), (3, 4))

# выведем теперь на печать как массив "a", так и комбинации его элементов
print('a:', a, sep='\n', end='\n\n')

print('элемент (0, 0):', a[0, 0])
print('элемент (2, 2):', a[2, 2])
print('строка 1: ', a[1, :])
print('столбец 2:', a[:, 2])

print('\nокно 2x2 (строки 1–2 и столбцы 0–1):', a[1:3, 0:2], sep='\n')


a:
[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]]

элемент (0, 0): 1.0
элемент (2, 2): 11.0
строка 1:  [5. 6. 7. 8.]
столбец 2: [ 3.  7. 11.]

окно 2x2 (строки 1–2 и столбцы 0–1):
[[ 5.  6.]
 [ 9. 10.]]


Разберем некоторые составляющие этого кода:

Команда `np.reshape()` (буквально, "поменяй форму") нужна для преобразования размерности получившегося массива `a`:
* первый аргумент функции – массив, размерность которого преобразуется.
* второй аргумент функции – кортеж, указывающий размерность нового массива, в который мы преобразуем исходный.

Команда `np.linspace(<начало>, <конец>, <количество>)` – это одномерная цепочка заданного *<количества>* равноотстоящих друг от друга вещественных чисел на промежутке от *<начала>* до *<конца>*.

В примере выше **преобразуемый** массив (первый аргумент) – это цепочка чисел от 1.0 до 12.0 *включительно* (ответьте самостоятельно, почему). Функция `np.reshape()` преобразует данную цепочку размерности (1, 12) – "1 строка и 12 столбцов" – в массив размерности (3, 4) – "3 строки и 4 столбца". Данный процесс происходит последовательно: первые 4 элемента цепочки попадают в первую строку, затем вторые 4 элемента – во вторую, и заключительные – в третью. Поскольку такое распределение арифметически возможно, ошибки не возникает.

Зато представленный ниже вариант не будет успешно выполнен. В массивах `NumPy` не может быть незаполненных ячеек в массиве заданного размера:

In [ ]:
np.reshape(np.linspace(1, 11, 11), (3, 4))


ValueError: cannot reshape array of size 11 into shape (3,4)

А теперь продемонстрируем более сложный функционал срезов на примере массива `b`:

In [ ]:
# перед запуском убедитесь, что ячейка [2] была выполнена

# добавим эти элементы как маркеры – чтобы можно было проверить себя
b[::-1, ::2, 2:3] = 12345
b[1:3, 0, 2] = 777

print('массив b')
print(b)
print('\n' + '-'*70 + '\n')
print('срез по первой оси', b[1:3, 0, 2], sep='\n', end='\n\n')
print('срез по трём осям', b[::-1, ::2, 2:3], sep='\n', end='\n\n')
print('срезы по первой и третьей осям', b[:, 2, :], sep='\n')

b.shape


массив b
[[[4.65570882e-310 0.00000000e+000 1.23450000e+004 6.94070458e-310
   6.94070458e-310]
  [6.94070458e-310 4.65570828e-310 6.94070463e-310 6.94070457e-310
   6.94070457e-310]
  [4.65570828e-310 4.65570828e-310 1.23450000e+004 4.65570828e-310
   4.65570828e-310]
  [6.94070457e-310 4.65570828e-310 6.94070457e-310 4.65570828e-310
   4.65570828e-310]]

 [[4.65570828e-310 4.65570828e-310 7.77000000e+002 4.65570828e-310
   4.65570828e-310]
  [4.65570828e-310 4.65570828e-310 4.65570828e-310 4.65570828e-310
   4.65570828e-310]
  [6.94070771e-310 6.94070771e-310 1.23450000e+004 9.59629402e-094
   1.30369751e-258]
  [1.22285193e-305 6.66508897e-186 6.47614143e-186 2.40141091e-253
   9.59629391e-094]]

 [[5.05888475e-104 4.89119613e-104 7.77000000e+002 4.09240880e-268
   2.40141091e-253]
  [1.16012076e-069 3.73707818e-214 1.74544696e-118 2.47731411e-253
   6.73128095e-239]
  [6.45716175e-186 9.40971091e-254 1.23450000e+004 6.12486990e+257
   4.18067950e+251]
  [9.07652381e+223 1.39806875e

(3, 4, 5)

Разберем, что мы получили.

1. `b[1:3, 0, 2]` звучит буквально так: из массива `b` возьми слои с первого (не нулевого) по третий не включительно, затем в каждом из них выбери нулевую строчку и второй столбец. Посмотрите на массив `b` и убедитесь, что результат верный.
2. `b[:, 2, :]` означает: возьми все слои, и в каждом выбери вторую строку и все столбцы.
3. Чуть более сложный вариант: `b[::-1, ::2, 2:3]` предлагаем вам озвучить самостоятельно. Вспомните, как работают срезы и что означает второе двоеточие.

Теперь давайте рассмотрим примеры обращения к элементам массива и их модификации (в т. ч. с использованием срезов).

In [ ]:
import numpy as np

a = np.ndarray((2, 5), dtype=float)      # создаём новый пробный массив
print('Исходный массив a', a, sep='\n')  # выведем его на печать

# занулим все элементы массива, стоящие на первой строке,
# а на второй – разместим единицы
a[0, :] = 0
a[1, :] = 1
print('\nЗанулили первую строку, по второй строке – единицы', a, sep='\n')

# сделаем каждый второй элемент первой строки равным 2
a[0, 1::2] = 2
print('\nКаждый второй элемент первой строки равен 2', a, sep='\n')

# перемешаем элементы строк массива в шахматном порядке
# без добавления ".copy()" результат будет не совсем таким, как мы ожидаем!!!
a[0, ::2], a[1, ::2] = a[1, ::2].copy(), a[0, ::2].copy()
print('\nПеремешаем элементы массива', a, sep='\n')

# используем выражение-генератор для изменения элементов 1 и 2 строк массива
a[0, 1::2] = [i+5 for i in range(len(a[0, :]))][1::2]
a[1, :]    = [i+1 for i in range(len(a[1, :]))]
print('\nЭлементы массива заданы выражениями-генераторами', a, sep='\n')


Исходный массив a
[[0.00000000e+000 0.00000000e+000 6.79038653e-313 6.79038653e-313
  6.79038653e-313]
 [1.93101617e-312 9.33678148e-313 1.14587773e-312 1.97345609e-312
  2.02566915e-322]]

Занулили первую строку, по второй строке – единицы
[[0. 0. 0. 0. 0.]
 [1. 1. 1. 1. 1.]]

Каждый второй элемент первой строки равен 2
[[0. 2. 0. 2. 0.]
 [1. 1. 1. 1. 1.]]

Перемешаем элементы массива
[[1. 2. 1. 2. 1.]
 [0. 1. 0. 1. 0.]]

Элементы массива заданы выражениями-генераторами
[[1. 6. 1. 8. 1.]
 [1. 2. 3. 4. 5.]]


### Специальные массивы:

Наряду с массивами общего типа, в модуле `NumPy` определены массивы специального типа. К ним относятся:
1. Нулевая матрица: `zeros(sh, dt)`;
2. Единичная матрица: `ones(sh, dt)`;
3. Единично-диагональная (1 на главной диагонали, 0 для других элементов):
    - `еуе(n, m)` для матрицы $n \times m$
    - `еуе(n)` или `identity(n)` – для квадратной матрицы.

В этих функциях:
* `sh` – **кортеж**, задающий размерность массива (shape),
* `dt` – параметр, определяющий тип данных (data type),
* `n`, `m` – некоторые натуральные числа.

Попробуем воспользоваться этими функциями.

In [ ]:
import numpy as np

a0 = np.zeros((2, 3), 'float')
print('a0:', a0, sep='\n')

a1 = np.ones((5, 8), 'int')
print('\na1:', a1, sep='\n')

a2 = np.eye(4, 6, dtype = int)
print('\na2:', a2, sep='\n')

a3 = np.identity (4, 'float')
print('\na3:', a3, sep='\n')


a0:
[[0. 0. 0.]
 [0. 0. 0.]]

a1:
[[1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]]

a2:
[[1 0 0 0 0 0]
 [0 1 0 0 0 0]
 [0 0 1 0 0 0]
 [0 0 0 1 0 0]]

a3:
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


### Математические функции

До сих пор мы рассматривали элементарные операции создания массива и обращения к его элементам. Наряду с этими операциями в модуле `NumPy` реализованы и всевозможные математические функции для работы с элементами массивов (избавляющие пользователя от необходимости последовательного обращения к каждому элементу массива для обработки содержащихся в них данных). 

Краткий (и отнюдь не исчерпывающий) список таких функций представлен ниже. Пусть `x` и `y` – некоторые `NumPy`-массивы. Тогда к ним применимы следующие действия:

|Функция |Описание: поэлементное... |Функция |Описание: возвращает поэлементно...|
|---|:-|:-|:-|
|`np.add(x, у)`|сложение двух массивов|`np.angle(x)`|углы комплексных чисел|
|`np.subtract(х, у)`|вычитание|`np.phase(x)`|фазы комплексных чисел|
|`np.multiply(x, у)`|умножение|`np.real(x)`|реальные части|
|`np.divide(x, у)`|деление|`np.imag(x)`|мнимые части|
|`np.power(x, у)`|возведение в степень|`np.conj(x)`|комплексно сопряженные|
|`np.negative(x, у)`|изменение знака в массиве `x` и запись в массив `y`|

Рассмотрим примеры:

In [ ]:
import numpy as np
import math

# единичный массив
a = np.ones((3, 4))

# создадим пробный массив, содержащий элементы ~pi
b = np.array([(i+1)*j*math.pi/12 for i in range(3) for j in range(4)])
# и переформатируем его под размерность "a"
b = np.reshape(b, (3, 4))
print('a =', a, sep='\n')
print('\nb =', b, sep='\n')

# сумма массивов a и b
c = np.add(a, b)
print('\na + b =', c, sep='\n')

# сумму можно задать и так
c = a + b
print('\na + b =', c, sep='\n')

# поэлементное возведение элементов массива "c"
# в степени из элементов массива "b"
d = np.power(c, b)
print('\n(a + b)^b =', d, sep='\n')

# массив, элементы которого есть sin(b_ij)
e = np.sin(b)
print('\nsin(b_ij) =', e, sep='\n')


a =
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]

b =
[[0.         0.26179939 0.52359878 0.78539816]
 [0.         0.52359878 1.04719755 1.57079633]
 [0.         0.78539816 1.57079633 2.35619449]]

a + b =
[[1.         1.26179939 1.52359878 1.78539816]
 [1.         1.52359878 2.04719755 2.57079633]
 [1.         1.78539816 2.57079633 3.35619449]]

a + b =
[[1.         1.26179939 1.52359878 1.78539816]
 [1.         1.52359878 2.04719755 2.57079633]
 [1.         1.78539816 2.57079633 3.35619449]]

(a + b)^b =
[[ 1.          1.06276979  1.24666805  1.57656642]
 [ 1.          1.24666805  2.11760879  4.40689589]
 [ 1.          1.57656642  4.40689589 17.33796361]]

sin(b_ij) =
[[0.         0.25881905 0.5        0.70710678]
 [0.         0.5        0.8660254  1.        ]
 [0.         0.70710678 1.         0.70710678]]


Обратите внимание на использование двойного генератора списка (list comprehension) в этом случае:
~~~python
b = np.array([(i+1)*j*math.pi/12 for i in range(3) for j in range(4)])
~~~

создает 3 строки (первый генератор, индекс `i`), в каждую из которых записывает 4 элемента (второй генератор, индекс `j`), определяемые текущими величинами `i` и `j`.

### Линейная алгебра

Функционал массивов `NumPy` естественным образом подходит для определения матриц линейной алгебры как самостоятельного типа данных. Напомним, что матрица в линейной алгебре – это некая таблица чисел размерностью $n \times m$. Реализация матриц в качестве отдельного типа данных в `NumPy` представляет частный, двумерный случай массива `ndarray`.

Для преобразования массива `NumPy` в матрицу используется команда `np.mat()`, например так:
~~~python
import numpy as np
a = np.ndarray(shape)  # некий кортеж – размерность массива "a"
a = np.mat(a)
~~~

Для матриц определены дополнительные (по сравнению с массивами) операции над элементами. Так, например, одной из основных операций линейной алгебры является **умножение матриц** (как частный случай, матрицы на вектор и вектора на вектор).

Для того, чтобы перемножить две матрицы, можно использовать следующие команды:

|Описание |Функция для объекта `ndarray` |Функция для объекта `matrix` |
|---|:-:|:-:|
|Поэлементное умножение|`*`|`multiply()`|
|Матричное умножение|`dot()`|`*`|

Рассмотрим примеры:

In [ ]:
import numpy as np

# создадим два массива
a = np.array([[1, 2], [3, 4]])
b = np.array([[5, 6], [7, 8]])
# и выведем их на печать
print('\nndarray – a :', a, sep='\n')
print('\nndarray – b :', b, sep='\n')

# разные виды умножений для массивов:
print('\nndarray – a*b :', a*b, sep='\n')
print('\nndarray – dot(a,b):', np.dot(a, b), sep='\n')

# преобразуем массивы в матрицы
c = np.mat(a)
d = np.mat(b)
print('\nmatrix - c :', c, sep='\n')
print('\nmatrix - d :', d, sep='\n')

# разные виды умножения для матриц:
print('\nmatrix - c*d :', c * d, sep='\n')
print('\nmatrix - multiply(a,b) :', np.multiply(c, d), sep='\n')



ndarray – a :
[[1 2]
 [3 4]]

ndarray – b :
[[5 6]
 [7 8]]

ndarray – a*b :
[[ 5 12]
 [21 32]]

ndarray – dot(a,b):
[[19 22]
 [43 50]]

matrix - c :
[[1 2]
 [3 4]]

matrix - d :
[[5 6]
 [7 8]]

matrix - c*d :
[[19 22]
 [43 50]]

matrix - multiply(a,b) :
[[ 5 12]
 [21 32]]


Отметим, что операция матричного произведения (независимо от типа перемножаемых массивов) может быть определена также при помощи оператора `@`:
~~~python
c = a @ b
~~~

In [ ]:
print(a @ b)
print()
print(c @ d)
print()
print(a@b == c@d)


[[19 22]
 [43 50]]

[[19 22]
 [43 50]]

[[ True  True]
 [ True  True]]


### Встроенные методы и функции для обработки массивов и матриц

Матрицы, как и массивы `NumPy`, реализуют ряд методов для обработки данных, которые в них хранятся. Ниже представлены некоторые из них.

В модуле `numpy` существует подмодуль линейной алгебры `linalg` – некоторые из приведенных в таблице функций прописаны именно в нем. В этом случае для вызова функции надо обращаться не к самой `np`, а через нее к подмодулю линейной алгебры: `np.linalg`. Подробнее об этом вы можете прочитать в [официальной документации](https://numpy.org/doc/stable/reference/routines.linalg.html).

|Сокращенная запись |Функция |Описание|
|:-:|:-:|:-|
|`.T`|`np.transpose()`|транспонирование с комплексным сопряжением|
|`.I`|`np.linalg.inv()`|обратная матрица|
|–|`np.mat()`|преобразование массива в матрицу|
|`.A`|–| преобразование матрицы в двумерный массив|
|–|`np.linalg.norm()`|вычисление нормы вектора|
|–|`np.linalg.cond()`|вычисление числа обусловленности матрицы|
|–|`np.linalg.det()`|вычисление определителя матрицы|
|–|`np.trace()`|след матрицы (сумма диагональных элементов)|

Еще раз напомним, что в `Python` заглавные и прописные буквы различимы! Не ошибитесь при наборе команд.

**N.B.**: Иногда для удобства (если в программе предстоит много работать с линейной алгеброй) подмодуль `linalg` импортируют отдельно под псевдонимом `LA`:
~~~python
from numpy import linalg as LA

print(LA.norm(a))
~~~

In [ ]:
import numpy as np

a = np.array([[1, 2], [3, 4]])

print('ndarray - a :', a, sep='\n')
print('\nnorm a:', np.linalg.norm(a))  # находим норму "a"
print('cond a:', np.linalg.cond(a))    # определяем число обусловленности "a"
print('det a:', np.linalg.det(a))      # детерминант "a"
print('trace a: ', np.trace(a))        # след массива "a"

b = np.mat(a)                          # преобразуем "a" в матрицу "b"
print('\ntype b: ', type(b))           # убедимся, что "b" – это матрица
print('type a.A: ', type(b.A))         # преобразуем обратно в массив


ndarray - a :
[[1 2]
 [3 4]]

norm a: 5.477225575051661
cond a: 14.933034373659265
det a: -2.0000000000000004
trace a:  5

type b:  <class 'numpy.matrix'>
type a.A:  <class 'numpy.ndarray'>


## Несколько заключительных слов

Перечисленные методы, конечно же, не исчерпывают всего множества возможностей, которые предоставляет в распоряжение исследователя модуль `NumPy`. Для полного списка функционала модуля рекомендуем обратиться к документации библиотеки, расположенной по [этому адресу](https://numpy.org/doc/1.23/user/index.html) (на момент подготовки материала лекции наиболее свежая версия `NumPy` – 1.23).

В ходе наших следующих занятий мы будем активно использовать массивы `NumPy` при решении задач. Применение специального инструментария языка `Python` для решения специфических задач, относящихся к области линейной алгебры, будет рассматриваться нами во второй половине курса.